scaletest/autostart/run.go

```diff
@@ -0,0 +1,206 @@
+		return xerrors.Errorf("create stop build: %w", err)
+	}
+
+	stopBuildCtx, cancel2 := context.WithTimeout(ctx, r.cfg.WorkspaceJobTimeout)
+	defer cancel2()
+
+	err = workspacebuild.WaitForBuild(stopBuildCtx, logs, newUserClient, stopBuild.ID)
+	if err != nil {
+		r.cfg.Metrics.AddError(newUser.Username, "wait_for_stop_build")
+		return xerrors.Errorf("wait for stop build to complete: %w", err)
+	}
+
+	logger.Info(ctx, fmt.Sprintf("workspace %q stopped successfully", workspace.Name))
+
+	logger.Info(ctx, "waiting for all runners to reach barrier")
+	r.cfg.SetupBarrier.Done()
+	r.cfg.SetupBarrier.Wait()
+	logger.Info(ctx, "all runners reached barrier, proceeding with autostart schedule")
+
+	autoStartTime := r.cfg.Clock.Now().Add(r.cfg.AutostartDelay)
+	schedule := fmt.Sprintf("CRON_TZ=UTC %d %d * * *", autoStartTime.Minute(), autoStartTime.Hour())
+
+	logger.Info(ctx, fmt.Sprintf("setting autostart schedule for workspace %q: %s", workspace.Name, schedule))
+
+	err = newUserClient.UpdateWorkspaceAutostart(ctx, workspace.ID, codersdk.UpdateWorkspaceAutostartRequest{
+		Schedule: &schedule,
+	})
+	if err != nil {
+		r.cfg.Metrics.AddError(newUser.Username, "update_workspace_autostart")
+		return xerrors.Errorf("update workspace autostart: %w", err)
+	}
+	close(r.autostartSet)
+
+	logger.Info(ctx, fmt.Sprintf("autostart schedule set for workspace %q", workspace.Name))
+
+	logger.Info(ctx, fmt.Sprintf("waiting for workspace %q to autostart", workspace.Name))
+
+	autostartInitiateCtx, cancel2 := context.WithTimeout(ctx, r.cfg.AutostartTimeout+r.cfg.AutostartDelay)
+	defer cancel2()
+
+	workspaceUpdates, err := newUserClient.WatchWorkspace(autostartInitiateCtx, workspace.ID)
+	if err != nil {
+		r.cfg.Metrics.AddError(newUser.Username, "watch_workspace")
+		return xerrors.Errorf("watch workspace: %w", err)
+	}
+
+	var autoStartBuild codersdk.WorkspaceBuild
+
+	logger.Info(ctx, "listening for workspace updates to detect autostart build")
+waitNewBuildLoop:
+	for {
+		select {
+		case <-autostartInitiateCtx.Done():
+			return xerrors.Errorf("timeout waiting for autostart build to be created: %w", autostartInitiateCtx.Err())
+		case updatedWorkspace, ok := <-workspaceUpdates:
+			if !ok {
+				r.cfg.Metrics.AddError(newUser.Username, "workspace_updates_channel_closed")
+				return xerrors.Errorf("workspace updates channel closed")
+			}
+
+			if updatedWorkspace.LatestBuild.ID != stopBuild.ID &&
+				updatedWorkspace.LatestBuild.Transition == codersdk.WorkspaceTransitionStart {
+				autoStartBuild = updatedWorkspace.LatestBuild
+				logger.Info(ctx, fmt.Sprintf("autostart build created with ID %s", autoStartBuild.ID))
+				break waitNewBuildLoop
+			}
+		}
+	}
+
+	logger.Info(ctx, "waiting for autostart build to complete")
+	buildCompleteCtx, cancel3 := context.WithTimeout(ctx, r.cfg.WorkspaceJobTimeout)
+	defer cancel3()
+
+	err = workspacebuild.WaitForBuild(buildCompleteCtx, logs, newUserClient, autoStartBuild.ID)
```

Comment:
I worry that polling for the build to complete here creates a bunch of load that could complicate analysis. Why not just continue to watch the `workspaceUpdates` until the build completes?